File: app/server/backend/routers/lab_result.py
"""孕程记 - 检查指标 API。"""

from typing import List, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import get_db
from backend.models.lab_result import LabResult
from backend.models.checkup import PrenatalCheckup
from backend.schemas.common import ApiResponse
from backend.schemas.lab_result import (
    LabResultCreate, LabResultUpdate, LabResultResponse, LabResultTrendResponse, LabResultTrendPoint,
)
from backend.services.reference_range import reference_range_service

router = APIRouter()


@router.post("/lab-results", response_model=ApiResponse[LabResultResponse])
async def create_lab_result(data: LabResultCreate, db: AsyncSession = Depends(get_db)):
    """录入检查指标，自动比对标准范围。"""
    # 自动比对标准范围
    ref_min, ref_max = data.reference_min, data.reference_max
    if ref_min is None or ref_max is None:
        # 获取产检记录的孕周
        checkup = await db.get(PrenatalCheckup, data.checkup_id)
        gest_week = checkup.gestational_week if checkup else None
        ref_range = reference_range_service.get_reference_range(data.category, data.item_name, gest_week)
        if ref_range:
            if ref_min is None:
                ref_min = ref_range[0]
            if ref_max is None:
                ref_max = ref_range[1]

    # 自动判断状态
    status = reference_range_service.evaluate_status(data.value, ref_min, ref_max)

    lab_result = LabResult(
        checkup_id=data.checkup_id,
        category=data.category,
        item_name=data.item_name,
        value=data.value,
        unit=data.unit,
        reference_min=ref_min,
        reference_max=ref_max,
        status=status,
    )
    db.add(lab_result)
    await db.flush()
    await db.refresh(lab_result)
    return ApiResponse(data=LabResultResponse.model_validate(lab_result))


@router.get("/lab-results", response_model=ApiResponse[List[LabResultResponse]])
async def list_lab_results(
    checkup_id: Optional[str] = None,
    category: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
):
    """指标列表，按 checkup_id 或 category 筛选。"""
    query = select(LabResult)
    if checkup_id:
        query = query.where(LabResult.checkup_id == checkup_id)
    if category:
        query = query.where(LabResult.category == category)
    result = await db.execute(query.order_by(LabResult.created_at))
    results = result.scalars().all()
    return ApiResponse(data=[LabResultResponse.model_validate(r) for r in results])


@router.get("/lab-results/trend", response_model=ApiResponse[LabResultTrendResponse])
async def get_lab_result_trend(
    category: str = Query(...),
    item_name: str = Query(...),
    pregnancy_id: str = Query(...),
    db: AsyncSession = Depends(get_db),
):
    """指标趋势（按 category + item_name）。"""
    query = (
        select(LabResult, PrenatalCheckup.gestational_week, PrenatalCheckup.checkup_date)
        .join(PrenatalCheckup, LabResult.checkup_id == PrenatalCheckup.id)
        .where(LabResult.category == category, LabResult.item_name == item_name)
        .where(PrenatalCheckup.pregnancy_id == pregnancy_id)
        .order_by(PrenatalCheckup.checkup_date)
    )
    result = await db.execute(query)
    rows = result.all()

    points = []
    unit = ""
    for row in rows:
        lab = row[0]
        unit = lab.unit
        points.append(LabResultTrendPoint(
            gestational_week=row[1],
            value=lab.value,
            checkup_date=row[2],
            reference_min=lab.reference_min,
            reference_max=lab.reference_max,
        ))

    return ApiResponse(data=LabResultTrendResponse(
        category=category, item_name=item_name, unit=unit, points=points,
    ))


@router.get("/lab-results/{lab_result_id}", response_model=ApiResponse[LabResultResponse])
async def get_lab_result(lab_result_id: str, db: AsyncSession = Depends(get_db)):
    """指标详情。"""
    lab = await db.get(LabResult, lab_result_id)
    if not lab:
        return ApiResponse(code=1001, data=None, message="指标不存在")
    return ApiResponse(data=LabResultResponse.model_validate(lab))


@router.put("/lab-results/{lab_result_id}", response_model=ApiResponse[LabResultResponse])
async def update_lab_result(lab_result_id: str, data: LabResultUpdate, db: AsyncSession = Depends(get_db)):
    """更新指标。"""
    lab = await db.get(LabResult, lab_result_id)
    if not lab:
        return ApiResponse(code=1001, data=None, message="指标不存在")
    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(lab, key, value)
    # 重新判断状态
    if data.value is not None:
        lab.status = reference_range_service.evaluate_status(lab.value, lab.reference_min, lab.reference_max)
    await db.flush()
    await db.refresh(lab)
    return ApiResponse(data=LabResultResponse.model_validate(lab))


@router.delete("/lab-results/{lab_result_id}", response_model=ApiResponse)
async def delete_lab_result(lab_result_id: str, db: AsyncSession = Depends(get_db)):
    """删除指标。"""
    lab = await db.get(LabResult, lab_result_id)
    if not lab:
        return ApiResponse(code=1001, data=None, message="指标不存在")
    await db.delete(lab)
    return ApiResponse(message="删除成功")

File: pregnancyjournal-0.0.26/pregnancyjournal-0.0.26/server/node/routes/lab_result.js
const express = require('express');
const router = express.Router();
const db = require('../db');
const config = require('../config');

function _determineStatus(value, reference_min, reference_max) {
  const numValue = parseFloat(value);
  if (reference_min !== null && reference_min !== undefined && numValue < parseFloat(reference_min)) return 'low';
  if (reference_max !== null && reference_max !== undefined && numValue > parseFloat(reference_max)) return 'high';
  return 'normal';
}

router.post('/lab-results', async (req, res) => {
  try {
    const { checkup_id, category, item_name, value, unit, reference_min, reference_max } = req.body;
    if (!checkup_id || !category || !item_name || value === undefined) {
      return res.json({ code: 1001, data: null, message: '缺少必填字段: checkup_id, category, item_name, value' });
    }
    const status = _determineStatus(value, reference_min, reference_max);
    const id = db.generateId();
    await db.run(
      `INSERT INTO lab_result (id, checkup_id, category, item_name, value, unit, reference_min, reference_max, status, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, datetime('now'), datetime('now'))`,
      [id, checkup_id, category, item_name, value, unit || null, reference_min || null, reference_max || null, status]
    );
    const row = await db.queryOne('SELECT * FROM lab_result WHERE id = ?', [id]);
    res.json({ code: 0, data: row, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/lab-results', async (req, res) => {
  try {
    const { checkup_id, category } = req.query;
    let where = 'WHERE 1=1';
    const params = [];
    if (checkup_id) { where += ' AND lr.checkup_id = ?'; params.push(checkup_id); }
    if (category) { where += ' AND lr.category = ?'; params.push(category); }
    const rows = await db.queryAll(
      `SELECT lr.* FROM lab_result lr ${where} ORDER BY lr.created_at DESC`,
      params
    );
    res.json({ code: 0, data: rows, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/lab-results/trend', async (req, res) => {
  try {
    const { category, item_name, pregnancy_id } = req.query;
    if (!category || !item_name || !pregnancy_id) {
      return res.json({ code: 1001, data: null, message: '缺少必填参数: category, item_name, pregnancy_id' });
    }
    const rows = await db.queryAll(
      `SELECT lr.*, pc.gestational_week, pc.checkup_date
       FROM lab_result lr
       JOIN prenatal_checkup pc ON lr.checkup_id = pc.id
       WHERE lr.category = ? AND lr.item_name = ? AND pc.pregnancy_id = ?
       ORDER BY pc.checkup_date ASC`,
      [category, item_name, pregnancy_id]
    );
    res.json({ code: 0, data: rows, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/lab_results/:id', async (req, res) => {
  try {
    const row = await db.queryOne('SELECT * FROM lab_result WHERE id = ?', [req.params.id]);
    if (!row) return res.json({ code: 1001, data: null, message: '记录不存在' });
    res.json({ code: 0, data: row, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.put('/lab_results/:id', async (req, res) => {
  try {
    const existing = await db.queryOne('SELECT id FROM lab_result WHERE id = ?', [req.params.id]);
    if (!existing) return res.json({ code: 1001, data: null, message: '记录不存在' });
    const fields = ['checkup_id', 'category', 'item_name', 'value', 'unit', 'reference_min', 'reference_max'];
    const sets = [];
    const params = [];
    for (const f of fields) {
      if (req.body[f] !== undefined) { sets.push(`${f} = ?`); params.push(req.body[f]); }
    }
    if (req.body.value !== undefined || req.body.reference_min !== undefined || req.body.reference_max !== undefined) {
      const current = await db.queryOne('SELECT value, reference_min, reference_max FROM lab_result WHERE id = ?', [req.params.id]);
      const val = req.body.value !== undefined ? req.body.value : current.value;
      const rMin = req.body.reference_min !== undefined ? req.body.reference_min : current.reference_min;
      const rMax = req.body.reference_max !== undefined ? req.body.reference_max : current.reference_max;
      sets.push(`status = ?`);
      params.push(_determineStatus(val, rMin, rMax));
    }
    sets.push("updated_at = datetime('now')");
    params.push(req.params.id);
    await db.run(`UPDATE lab_result SET ${sets.join(', ')} WHERE id = ?`, params);
    const row = await db.queryOne('SELECT * FROM lab_result WHERE id = ?', [req.params.id]);
    res.json({ code: 0, data: row, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.delete('/lab_results/:id', async (req, res) => {
  try {
    const existing = await db.queryOne('SELECT id FROM lab_result WHERE id = ?', [req.params.id]);
    if (!existing) return res.json({ code: 1001, data: null, message: '记录不存在' });
    await db.run('DELETE FROM lab_result WHERE id = ?', [req.params.id]);
    res.json({ code: 0, data: null, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

module.exports = router;
